compiler/rustc_infer/src/infer/error_reporting/suggest.rs

```diff
@@ -351,6 +351,70 @@ impl<'tcx> TypeErrCtxt<'_, 'tcx> {
         }
     }
 
+    pub(super) fn suggest_function_pointers(
+        &self,
+        cause: &ObligationCause<'tcx>,
+        span: Span,
+        exp_found: &ty::error::ExpectedFound<Ty<'tcx>>,
+        diag: &mut Diagnostic,
+    ) {
+        debug!("suggest_function_pointers(cause={:?}, exp_found={:?})", cause, exp_found);
+        let ty::error::ExpectedFound { expected, found } = exp_found;
+        let expected_inner = expected.peel_refs();
+        let found_inner = found.peel_refs();
+        if !expected_inner.is_fn() || !found_inner.is_fn() {
+            return;
+        }
+        match (&expected_inner.kind(), &found_inner.kind()) {
+            (ty::FnPtr(sig), ty::FnDef(did, substs)) => {
+                let sig = &(self.normalize_fn_sig)(*sig);
+                let fn_name = self.tcx.def_path_str_with_substs(*did, substs);
+                if !self.same_type_modulo_infer(expected_inner, found_inner)
```

Comment:
How about this:

```suggestion
            (ty::FnPtr(sig), ty::FnDef(did, substs)) => {
                let expected_sig = &(self.normalize_fn_sig)(*sig);
                let found_sig = &(self.normalize_fn_sig)(self.tcx.bound_fn_sig(did).subst(tcx, substs));
                let fn_name = self.tcx.def_path_str_with_substs(*did, substs);
                if !self.same_type_modulo_infer(found_sig, expected_sig)
```

And then change `same_type_modulo_infer` from taking two `Ty` to taking two [`T: Relate`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/ty/relate/trait.Relate.html) so we can compare the sigs themselves.

Reply:
It worked! The majority of affected tests have better suggestions now.